crates/holochain_sqlite/src/db/p2p_agent_store.rs

```diff
@@ -102,6 +102,105 @@ impl AsP2pAgentStoreConExt for crate::db::PConnGuard {
     }
 }
 
+use once_cell::sync::Lazy;
+use parking_lot::Mutex;
+use std::collections::{hash_map, HashMap};
+
+#[allow(clippy::type_complexity)]
+static AI_CACHE: Lazy<Mutex<HashMap<String, HashMap<Arc<KitsuneAgent>, AgentInfoSigned>>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+fn ai_cache_load(con: &Connection) -> DatabaseResult<HashMap<Arc<KitsuneAgent>, AgentInfoSigned>> {
+    let mut stmt = con
+        .prepare(sql_p2p_agent_store::SELECT_ALL)
+        .map_err(|e| rusqlite::Error::ToSqlConversionFailure(e.into()))?;
+    let mut out = HashMap::new();
+    for r in stmt.query_map([], |r| {
```

Comment:
isn't the one `r` shadowing the closure's parameter `r` here? If so, I prefer separate identifiers

Reply:
I thought that and had to check it out and tweak it to understand, because the output of the match is an `&mut HashMap` you have a reference to the lock via the element reference that can't be dropped until you release the element reference. It's very subtle but I believe this is correct

Reply:
This is a copy-paste of another query below. But I can change the idents : )

Reply:
Ahh ignore me, I was looking at the `ai_cache_lock ` macro

Reply:
Yeah, not super happy with that macro, but there isn't a good way in rust to return both a binding and a thing referencing that binding, so had to be a macro...

Reply:
other than [ouroboros](https://crates.io/crates/ouroboros)